src/zarr/codecs/batched_codec_pipeline.py

```diff
@@ -0,0 +1,439 @@
+        from zarr.codecs.sharding import ShardingCodec
+
+        if not any(isinstance(codec, ArrayBytesCodec) for codec in codecs):
+            raise ValueError("Exactly one array-to-bytes codec is required.")
+
+        prev_codec: Codec | None = None
+        for codec in codecs:
+            if prev_codec is not None:
+                if isinstance(codec, ArrayBytesCodec) and isinstance(prev_codec, ArrayBytesCodec):
+                    raise ValueError(
+                        f"ArrayBytesCodec '{type(codec)}' cannot follow after "
+                        + f"ArrayBytesCodec '{type(prev_codec)}' because exactly "
+                        + "1 ArrayBytesCodec is allowed."
+                    )
+                if isinstance(codec, ArrayBytesCodec) and isinstance(prev_codec, BytesBytesCodec):
+                    raise ValueError(
+                        f"ArrayBytesCodec '{type(codec)}' cannot follow after "
+                        + f"BytesBytesCodec '{type(prev_codec)}'."
+                    )
+                if isinstance(codec, ArrayArrayCodec) and isinstance(prev_codec, ArrayBytesCodec):
+                    raise ValueError(
+                        f"ArrayArrayCodec '{type(codec)}' cannot follow after "
+                        + f"ArrayBytesCodec '{type(prev_codec)}'."
+                    )
+                if isinstance(codec, ArrayArrayCodec) and isinstance(prev_codec, BytesBytesCodec):
+                    raise ValueError(
+                        f"ArrayArrayCodec '{type(codec)}' cannot follow after "
+                        + f"BytesBytesCodec '{type(prev_codec)}'."
+                    )
+            prev_codec = codec
+
+        if any(isinstance(codec, ShardingCodec) for codec in codecs) and len(codecs) > 1:
+            warn(
+                "Combining a `sharding_indexed` codec disables partial reads and "
+                + "writes, which may lead to inefficient performance."
+            )
+
+        return (
+            tuple(codec for codec in codecs if isinstance(codec, ArrayArrayCodec)),
+            next(codec for codec in codecs if isinstance(codec, ArrayBytesCodec)),
+            tuple(codec for codec in codecs if isinstance(codec, BytesBytesCodec)),
+        )
+
+    @classmethod
+    def from_list(cls, codecs: list[Codec], *, batch_size: int | None = None) -> Self:
+        array_array_codecs, array_bytes_codec, bytes_bytes_codecs = cls.codecs_from_list(codecs)
+
+        return cls(
+            array_array_codecs=array_array_codecs,
+            array_bytes_codec=array_bytes_codec,
+            bytes_bytes_codecs=bytes_bytes_codecs,
+            batch_size=batch_size or config.get("codec_pipeline.batch_size"),
+        )
+
+    @property
+    def supports_partial_decode(self) -> bool:
+        return (len(self.array_array_codecs) + len(self.bytes_bytes_codecs)) == 0 and isinstance(
+            self.array_bytes_codec, ArrayBytesCodecPartialDecodeMixin
+        )
```

Comment:
let's add a docstring or comment here. The comment should answer why having an array_array_codec or bytes_bytes_codec in the pipeline means `supports_partial_decode is False`.